allennlp/data/dataset_readers/reading_comprehension/triviaqa.py

```diff
@@ -1,32 +1,280 @@
+
+
+class Question(NamedTuple):
+    """
+    We process each TriviaQA question into tokenized question text
+    and merged paragraphs that may or may not contain the answer.
+    """
+    id: str  # pylint: disable=invalid-name
+    text: str
+    tokens: List[Token]
+    paragraphs: MergedParagraphs
+    answer_texts: List[str] = None
+
+    def to_json(self) -> JsonDict:
+        return {
+                "id": self.id,
+                "text": self.text,
+                "tokens": [token_to_json(token) for token in self.tokens],
+                "paragraphs": self.paragraphs.to_json(),
+                "answer_texts": self.answer_texts
+        }
+
+    @staticmethod
+    def from_json(blob: JsonDict) -> 'Question':
+        return Question(id=blob['id'],
+                        text=blob['text'],
+                        tokens=[json_to_token(token) for token in blob['tokens']],
+                        paragraphs=MergedParagraphs.from_json(blob['paragraphs']),
+                        answer_texts=blob.get('answer_texts'))
+
+
+def _merge(paragraphs: List[str],
+           question: str,
+           answer_texts: List[str],
+           tokenizer: Tokenizer,
+           max_paragraphs: int = 4,
+           tfidf_sort: bool = True,
+           require_answer: bool = False,
+           max_paragraph_length: int = 400) -> MergedParagraphs:
+    """
+    Given some (small?) paragraphs and a question, merge the paragraphs to make
+    synthetic paragraphs of size at most ``max_paragraph_length``, and return a ``MergedParagraphs``
+    instance containing the "best" ``topn`` of them. If ``topn`` is None, return all of them.
+
+    If ``tfidf_sort`` is True, "best" means closest (in terms of tfidf distance) to the question.
+    If ``require_answer`` is True, then the best "best" paragraph is also required to contain
+    one of the ``answer_texts``. Don't do this with a test or validation dataset, that's cheating.
+    If neither is True, then just the first ``topn`` paragraphs are returned.
+    """
+    merged_paragraphs: List[str] = []
+
+    current_length = 0
+    current_paragraph = ""
+
+    for paragraph in paragraphs:
+        paragraph_tokens = tokenizer.tokenize(paragraph)
+        paragraph_length = len(paragraph_tokens)
+
+        too_long = current_length + paragraph_length > max_paragraph_length
+
+        if too_long and current_length > 0 and paragraph_length <= max_paragraph_length:
+            # Keep the previous paragraph and start a new one
+            merged_paragraphs.append(current_paragraph)
+            current_length = paragraph_length
+            current_paragraph = paragraph
+
+        elif too_long and current_length > 0 and paragraph_length > max_paragraph_length:
+            # Keep the previous paragraph and truncate this one
+            truncated = ' '.join(token.text for token in paragraph_tokens[:max_paragraph_length])
+
+            merged_paragraphs.append(current_paragraph)
```

Comment:
Why are we adding the current_paragraph to the merged_paragraphs? 
Does it not make sense to just add a truncated version of the long paragraph to the merged_paragraph and do nothing with the current_paragraph. 
This way the current_paragraph can be accumulated with later paragraphs which are also small?

Reply:
`current_paragraph` is what we have already from previous iterations of the loop.  In my first read of this code, I was also confused by this, as I just jumped to line 118 and assumed that `current_length` was the length of the paragraph at the current iteration.  That's actually `paragraph_length`.

Reply:
I think there is a misunderstanding between the two of us. 
So I have written a test case to explain my point. 
In the test case there are 3 paragraphs. 
The first and the last are very short, with the middle one being longer than the max allowed. 


    def test_merge(self):
        short_paragraph_1 = "short paragraph before the long paragraph"
        really_long_para = "Paul is 28 years old. this is a really long para which is much longer than the max length of a paragraph we expect this to be truncated"
        short_paragraph_2 = "short paragraph after the long paragraph"
        paragraphs = [short_paragraph_1, really_long_para, short_paragraph_2]

        re = _merge(paragraphs=paragraphs,
                    question="what age is Paul",
                    answer_texts='28',
                    tokenizer=WordTokenizer.from_params(Params({})),
                    max_paragraphs=4,
                    tfidf_sort=True,
                    require_answer=False,
                    max_paragraph_length=16)

        assert len(re.texts) == /* what should the number here be? */


At the moment the MergedParagraph returned from the _merge method consists of 3 texts:
0 = {str} ' @@PARAGRAPH@@ short paragraph before the long paragraph'
1 = {str} 'Paul is 28 years old . this is a really long para which is much longer'
2 = {str} ' @@PARAGRAPH@@ short paragraph after the long paragraph'

However I would expect it to return 2 texts: 
0 = {str} 'Paul is 28 years old . this is a really long para which is much longer'
1 = {str} ' @@PARAGRAPH@@ short paragraph before the long paragraph @@PARAGRAPH@@ short paragraph after the long paragraph'

Do you agree with this?